internal/function/diff.go

```diff
@@ -0,0 +1,262 @@
+package function
+
+import (
+	"context"
+	"fmt"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	fmtdiff "gopkg.in/src-d/go-git.v4/plumbing/format/diff"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+
+	"github.com/src-d/gitbase"
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+)
+
+type (
+	// Diff implements git-diff function as a UDF
+	Diff struct {
+		repo sql.Expression
+		from sql.Expression
+		to   sql.Expression
+	}
+
+	// diffItem stands for a single payload item
+	// (response payload is an array of JSON objects as strings)
+	// Indexes are local per file.
+	// Many objects for different files may have the same index.
+	diffItem struct {
+		Index  int    `json:"index"`
+		File   string `json:"file"`
+		Chunk  string `json:"chunk"`
+		Status string `json:"status"`
+	}
+)
+
+// NewDiff constructor
+func NewDiff(args ...sql.Expression) (sql.Expression, error) {
+	d := &Diff{}
+	switch len(args) {
+	case 2:
+		d.repo, d.to = args[0], args[1]
```

Comment:
Unless I am wrong, you can recover the repo name from `row[0]` as is done in the squash and works for me at: https://github.com/mcuadros/gitbase/commit/39ca34d520350c6089181591400d1056fa05c778#diff-eaff2b7fc045a1ea059d7a8d72e864ccR89

Reply:
That's only pure luck. You can never rely on the contents of the row to get the repository.

If you do this it won't work:

```sql
SELECT commit_stats(arguments) FROM (SELECT all args except repository id FROM tables) t
```

Reply:
If can't be done in this way, we should provide the ability to recover the repositories without the requirement of passing the repository.

/cc @smola

Reply:
If can't be done in this way, we should provide the ability to recover the repositories without the requirement of passing the repository.

/cc @smola

Reply:
Let's take a step back - `diff` is a function, so should work without any global context, independent. In other words diff(hash1, hash2) means nothing. You have to specify repo.
So why to take a repo from the different source than commits.
Either take all from args or from rows, but mix will be harder to test. 
Current implementation lets you just say `SELECT diff("myrepo", "hash1", "hash2")`
without repo it will require some data, under the hood. 
Last but not least, with select (without repo) like
`diff(hash1, hash2)` you don't know what was tested original repo, or maybe fork.